### PR TITLE
Lao translations AND dynamic configuration of language menu

### PR DESCRIPTION
**Pull request by Abby-Wheelis**
This PR adds a lao.json file to the i18n options so Lao can be more supported on the join page. 

The other change here, in line with expanding languages, is having the language menu configure dynamically, based on the translated text in the config file. This aligns with the conversation in [issue #931](https://github.com/e-mission/e-mission-docs/issues/931). A brief overview of the code:

- list of supported languages is compiled after the config file loads
- this list creates a preliminary menu (with the i18n codes as the label text)
- a mapping of language codes to their names in that language is created once the translation files load
- the menu is re-loaded

When I've tried loading the menu just once, I end up with extra options, but loading once with the default texts and then again seems to prevent that. Looking at my log statements in the console, I can see that there is a timing factor with the two batched promises loading, so the extra menu load might just be convenient timing.

In summary: it all works, but I am not confident in the timing of the different menu configuration calls, so I would appreciate a second set of eyes and feedback on that aspect!

### CODE REVIEW

File: frontend/index.html [hunk 1]
```diff
@@ -759,6 +828,8 @@ <h2 data-i18n="contact">Contact</h2>
             } else {
                 hideTokenAndURL();
             }
+            // only place I've been able to ensure the right data exists before proceeding
+            SUPPORTED_LANGUAGES = getSupportedLanguages();
```
Comment (Abby-Wheelis) [inline frontend/index.html:832]: This is the best place I've found to load the supported languages, since the data from the config file will exist here

File: frontend/index.html [hunk 2]
```diff
@@ -783,7 +856,11 @@ <h2 data-i18n="contact">Contact</h2>
         // https://stackoverflow.com/questions/1414365/disable-enable-an-input-with-jquery
         Promise.all(otherI18nLoadPromises).then(() => {
             console.log("Finished loading other languages "+toLoadLanguages);
-            $('#langSel').prop('disabled', false)
+            //now that all i18n have been loaded, get map
+            LANG_TO_NAME = makeKeyNameMap(SUPPORTED_LANGUAGES);
```
Comment (Abby-Wheelis) [inline frontend/index.html:860]: This is where I've determined to load the code to name mapping and create the menu, since the i18n information exists here

**Review by Abby-Wheelis — COMMENTED** [review]
@shankari it looks like from the Blame that you've done the most recent work with these promises - do you have any suggestions for weaving in the language menu configuration? I'm struggling with having the correct subset of languages and the actual names of the languages since the config and i18n files are loaded in different places / on different timelines

File: frontend/index.html [hunk 3]
```diff
@@ -725,16 +792,18 @@ <h2 data-i18n="contact">Contact</h2>
...
         // want to redisplay the text twice
         // https://stackoverflow.com/a/16979926/4040267
+        
         $('#langSel').val(curr_lang);
-        const SUPPORTED_LANGUAGES = $.map($('#langSel')[0].options, (e) => e.value);
-        console.log("After reading from options, languages are ", SUPPORTED_LANGUAGES);
+        //assuming all languages in i18n folder are possible
+        //used to be all the langs from the menu, but that's dynamic now, and not loaded yet
+        SUPPORTED_LANGUAGES = ["en", "es", "lo"];
+        console.log("Based on the list, languages are ", SUPPORTED_LANGUAGES);
```
Comment (shankari) [inline frontend/index.html:806]: I would remove this. You are correct that at this point, we don't have the list of languages. You can just have the dropdown disabled by default until we have read in the languages.

File: frontend/index.html (hunk 2 above)
Comment (shankari) [reply frontend/index.html:860]: Yup, this is correct.

**Review by shankari — CHANGES_REQUESTED** [review]
High level comments below.
Also, make sure that you are super familiar with javascript promises and async/await.
We use them a lot everywhere.
Take the time to work through some tutorials in case you have not done so already

File: frontend/index.html (hunk 3 above)
Comment (Abby-Wheelis) [reply frontend/index.html]: removed this in [78c9c78](https://github.com/e-mission/nrel-openpath-join-page/pull/18/commits/78c9c786025c4398e4f15174b8ca4ac914441c0e)

File: frontend/i18n/en.json [hunk 4]
```diff
@@ -1,4 +1,5 @@
 {
+	"_language_name": "English",
```
[the remark this answers is not in the document]
Comment (Abby-Wheelis) [reply frontend/i18n/en.json]: I added it into the metadata, but I'm having a difficult time accessing it later `$.i18n('@metadata.language_name')` does not work like `$.i18n("language_name") was -- I found an escape method but printing `$.i18n(- '@metadata.language_name')` to the console makes it seem like each i18n has it's own metadata? 
Screenshot 2023-07-07 at 3 46 21 PM

[the remark this answers is not in the document]
Comment (shankari) [reply frontend/i18n/en.json]: I am fine with leaving it outside for now in that case.
We should be able to look at the `i18n` code (it is open source) to see why `@metadata` doesn't work (and if it is supposed to work) but not sure we have the time for it now.

File: frontend/index.html (hunk 1 above)
Comment (Abby-Wheelis) [reply frontend/index.html:828]: The issue I've been having is that since the supported languages come from the dynamic config, we don't know which ones to load until loading the dynamic config file is loaded, so I think we need to wait for the dynamic config to be done loading before we load any languages -- this will allow for configuring the default too!

I am generally familiar with what promises are in Javascript, but haven't worked with `async` and `await` yet, I'm looking into those now and plan to try using `await` for dynamic config since all language info not depends on that.

Comment (shankari) [reply frontend/index.html:828]: > The issue I've been having is that since the supported languages come from the dynamic config, we don't know which ones to load until loading the dynamic config file is loaded, so I think we need to wait for the dynamic config to be done loading before we load any languages -- this will allow for configuring the default too!

Correct. Now all promises need to be moved into the `then` for the dynamic load.

> I'm looking into those now and plan to try using await for dynamic config since all language info now depends on that.

Yup! Note that you can only use `await` in an `async` function or a top-level module. Not sure if `document.ready()` qualifies, but it seems like it should be the toppest of all top modules 😄

Comment (shankari) [reply frontend/index.html:828]: https://stackoverflow.com/questions/69984178/async-await-with-jquery-document-ready

Comment (Abby-Wheelis) [reply frontend/index.html]: I just reconfigured, but there's still a bug somewhere! I don't have time to find it right now, but will tonight or tomorrow! Everything loads and displays in the right languages at the right times (behaving as expected) with the exception of the `language_name` where it's defaulting to English - I find it odd that all the other keys load fine and the page is 99% translated, but I'll look into it when I get a moment

Comment (Abby-Wheelis) [reply frontend/index.html]: I wrapped up the reconfiguring of the load sequences -- now it loads the dynamic config with `await`, then the default (first in config) i18n file with `await`, and finally loads the other language(s) with a traditional promise. In the `then` for the other language loads, the language menu is set up and enabled. I tried using `await` for all three load processes, but that was yielding the issue from yesterday - this way works consistently.

I also added the `dashboard-preview-lo.png` which I noticed was missing in debugging.

File: frontend/index.html [hunk 5]
```diff
@@ -722,53 +787,57 @@ <h2 data-i18n="contact">Contact</h2>
...
-        // END: First round of loading
-
-        // BEGIN: Second round of loading
-        // then, we load all the other languages
+        const defaultI18nLoadPromise = $.i18n().load('i18n/'+default_language+'.json', default_language);
+        console.log("Default i18n Promise", defaultI18nLoadPromise); //supposed to be undefined? works and looks undefined in the original code?
+        const i18nLoadResult = await defaultI18nLoadPromise;
+        console.log("Default i18n result is", i18nLoadResult);
+
+        $('#langSel').val(default_language);
```
Comment (shankari) [inline frontend/index.html:835]: For my own edification, `#langSel` won't actually have any entries here, right? (that will happen in `makeLanguageMenu` after the other promises are loaded).

So I don't think this will actually do anything. It doesn't hurt anything, so I am OK to merge now, but you might want to test that out to understand the code better and polish it further.

Unless you think that the `val` is stored somehow and applied when the values are added in which case, document that for the next person working on this code.

**Review by shankari — APPROVED** [review]

**Comment by shankari** [comment]
I don't see a lot of blame churn here, so I am going to merge with squashing.
Congrats! @Abby-Wheelis

File: frontend/index.html (hunk 5 above)
Comment (Abby-Wheelis) [reply frontend/index.html:835]: You were absolutely right, I moved this line of code until after the language is created and enabled (after the other promises are loaded). This lets the name of the language still be displayed in the menu as soon as the menu is ready! The change is reflected in #19, which can address these and any other polishing changes we decide to make.
